### DISCUSSION TITLE
Duck typing via anonymous interface implementation wrappers

### DISCUSSION THREAD

**Discussion by kellybirr** [dotnet/csharplang#5793] [Language Ideas] [answered]
I would propose that a form of "duck typing" to interfaces could be added to a future version of the C# language and compiler by allowing the use of `new` to create an instance of an interface with a constructor-like syntax that takes a single object as the parameter.

Exmaple"
```csharp
// an interface
interface IDuck {
  void Walk(int steps);
  void Quack(double volume);
}

// a class that does not implement the interface but matches the pattern.
class PersonActingLikeADuck {
  public void Walk(int steps) { 
    // do the thing
  }
  void Quack(double volume) {
    // make your best duck sounds
  }
}

// create a person class
var person = new PersonActingLikeADuck();

// create an IDuck from the person
var duck = new IDuck(person);

// do things
duck.Walk(10);
duck.Quack(0.90);

```

The compiler could verify that the class meets the contract defined by the interface and generate an anonymous class that actually implements the interface and wraps the provided instance of the class, passing through all the methods and properties.

Also, this could be combined with anonymous classes as follows:

```csharp
var anotherDuck = new IDuck(new {
  Walk = delegate(int steps) { /* ... */ },
  Quack = delegate(double volume) { /* ... */ }
});
```

**Comment by jnm2** [comment]
This was discussed at https://github.com/dotnet/roslyn/issues/13#issuecomment-173413281 and https://github.com/dotnet/csharplang/discussions/1542.

**Reply by kellybirr** [reply]
It does look very similar in style and intent.  I'm just proposing a slightly different syntax and implementation method. that solves some of the concerns raised in previous discussions.

**Reply by kellybirr** [reply]
I'm guessing that will probably be the answer again, but it doesn't hurt to bring it up with this proposed syntax, years later.  There are many exciting things happening in C# these days.  This, or something like it, would be extremely useful, when combined with existing extension methods that .net 5 and 6 make heavy use of.

**Reply by jnm2** [reply]
Also, the `new IDuck(person)` concept is similar to https://github.com/dotnet/csharplang/issues/5497.

**Comment by hez2010** [comment]
This would cause some interfaces being implemented accidentally by undesired types.  

Instead of duck typing, we do have more mature and safer solution to extensibility, that is trait/shape.

```csharp
// an interface
interface IDuck {
  void Walk(int steps);
  void Quack(double volume);
}

// a class that does not implement the interface.
class PersonActingLikeADuck {
}

// extend PersonActingLikeADuck to implement IDuck
extension MyDuckExtension : IDuck, PersonActingLikeADuck
{
  public void Walk(int steps) { 
    // do the thing
  }
  void Quack(double volume) {
    // make your best duck sounds
  }
}

// create a person class
var person = new PersonActingLikeADuck();

// the person already implements IDuck so we can assign it to IDuck directly
IDuck duck = person;

// do things
duck.Walk(10);
duck.Quack(0.90);
```

**Reply by kellybirr** [reply]
That's cool

**Reply by kellybirr** [reply]
Question, how would requiring the `new IDuck(object)` syntax cause undesired implementations?  I'm not so much arguing as I don't see it.  It's part of the reason for the explicit syntax requirement in my proposal as opposed to a more natural duck typing that would be inferred.

**Comment by 333fred** [comment, marked as answer]
`new IDuck()` is already legal syntax and (unfortunately) means something. Let me introduce you to COM, and the ability to instantiate an interface:

```cs
// Assembly 1
using System;
using System.Runtime.InteropServices;
using System.Runtime.CompilerServices;
[assembly: PrimaryInteropAssemblyAttribute(1,1)]
[assembly: Guid(""f9c2d51d-4f44-45f0-9eda-c9d599b58257"")]
[ComImport()]
[Guid(""f9c2d51d-4f44-45f0-9eda-c9d599b58279"")]
[CoClass(typeof(ClassITest33))]
public interface ITest33 : System.Collections.IEnumerable
{
    void Add(int x);
}
[Guid(""f9c2d51d-4f44-45f0-9eda-c9d599b58278"")]
public abstract class ClassITest33
{
}
```

```cs
// Assembly 2, references Assembly 1
class UsePia
{
    public void M1()
    {
	    ITest33 x = new ITest33();
    }
}
```

**Reply by jnm2** [reply]
@333fred This could light up in places where `new IFoo` fails due to not being a COM interface.

**Reply by HaloFour** [reply]
It's always felt weird to me that C# has some shadow features that only work when using COM interop.

**Reply by 333fred** [reply]
> @333fred This could light up in places where `new IFoo` fails due to not being a COM interface.

I mean, it _could_, but that feels like the type of ambiguity I want no part of (aside from me not liking duck typing in the first place).

**Reply by kellybirr** [reply]
I totally forgot about COM interop!  I suppose it's a good thing that I haven't used it in long enough to actually forget about it.

I'm not sure I never knew this was legal C# for COM Interfaces.  

Thanks for helping me understand.!

**Reply by jnm2** [reply]
@333fred `new Foo` could be a COM class call. Why would `new IFoo` be more ambiguous than that?

**Reply by bernd5** [reply]
Is that somewhere specified in this repo?
